src/wodan-irmin/wodan_irmin.ml

```diff
@@ -274,36 +289,37 @@ functor
 
     let () = assert (K.hash_size = DB.Stor.P.key_size)
 
+    let val_to_inner va =
+      let raw_v = Irmin.Type.to_bin_string V.t va in
+      let k = K.hash (fun f -> f raw_v) in
+      let raw_k = Irmin.Type.to_bin_string K.t k in
+      (k, DB.Stor.key_of_string raw_k, DB.Stor.value_of_string ("C" ^ raw_v))
```

Comment:
```suggestion
    let value_to_string = Irmin.Type.to_bin_string V.t
    let key_to_string = Irmin.Type.to_bin_string K.t
    
    let val_to_inner va =
      let raw_v = value_to_string va in
      let k = K.hash (fun f -> f raw_v) in
      let raw_k = key_to_string k in
      (k, DB.Stor.key_of_string raw_k, DB.Stor.value_of_string ("C" ^ raw_v))
```
Specialising the generic operations to a type representation may be an expensive operation (and definitely allocates more than you want.) It's worth hoisting the specialisation out of your hot loops to reduce minor heap churn.

Reply:
Thanks! I have no idea how much the optimizer does vs how much I should help it, it's worth a look.

I'll pick up your suggestion in a separate PR, and see if I can pick up other similar specialisations that could be moved out.